tools/keytools/keygen.c

```diff
@@ -491,13 +492,54 @@ static void key_import(uint32_t ktype, const char *fname)
     uint8_t buf[KEYSLOT_MAX_PUBKEY_SIZE];
     FILE *f;
     int r;
+    int i;
+    int keySize = 0;
     f = fopen(fname, "rb");
     if (f == NULL) {
         fprintf(stderr, "Fatal error: could not open file %s to import public key\n", fname);
         exit(6);
     }
-    r = fread(buf, sizeof(buf), 1, f);
-    keystore_add(ktype, buf, r, fname);
+
+    switch (ktype) {
+        case KEYGEN_ED25519:
+            keySize = KEYSTORE_PUBKEY_SIZE_ED25519;
+            break;
+        case KEYGEN_ED448:
+            keySize = KEYSTORE_PUBKEY_SIZE_ED448;
+            break;
+        case KEYGEN_ECC256:
+            keySize = KEYSTORE_PUBKEY_SIZE_ECC256;
+            break;
+        case KEYGEN_ECC384:
+            keySize = KEYSTORE_PUBKEY_SIZE_ECC384;
+            break;
+        case KEYGEN_RSA2048:
+            keySize = KEYSTORE_PUBKEY_SIZE_RSA2048;
+            break;
+        case KEYGEN_RSA3072:
+            keySize = KEYSTORE_PUBKEY_SIZE_RSA3072;
+            break;
+        case KEYGEN_RSA4096:
+            keySize = KEYSTORE_PUBKEY_SIZE_RSA4096;
+            break;
+        default:
+            keySize = 0;
+    }
+
+    i = 0;
+
+    do
+    {
+        r = fread(buf + i, 1, 1, f);
+        i += r;
+    } while ( i < keySize && r > 0);
+
+    if (i <= 0 || i > keySize) {
```

Comment:
Should `r` also be checked for `< 0`?

Reply:
yes, I was thinking i would wind up negative but not if it fails after reading a few bytes